File: go/hearth/bech32m.go
package hearth

import "strings"

// Bech32m encoder/decoder (BIP-350), checksum constant 0x2bc830a3.

const (
	bech32mCharset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
	bech32mConst   = 0x2bc830a3
)

var bech32mGen = [5]int{0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3}

func Bech32mEncode(hrp string, data []byte) string {
	values := convertBits(bytesToInts(data), 8, 5, true)
	if values == nil {
		panic("cannot regroup payload bits")
	}
	values = append(values, bech32mCreateChecksum(hrp, values)...)
	var sb strings.Builder
	sb.WriteString(hrp)
	sb.WriteByte('1')
	for _, d := range values {
		sb.WriteByte(bech32mCharset[d])
	}
	return sb.String()
}

// Bech32mDecodeRaw verifies structure + checksum only, returning the hrp and the
// raw 5-bit data groups (checksum stripped). ok=false on any error.
func Bech32mDecodeRaw(s string) (hrp string, data []int, ok bool) {
	if s != strings.ToLower(s) && s != strings.ToUpper(s) {
		return "", nil, false
	}
	lower := strings.ToLower(s)
	pos := strings.LastIndex(lower, "1")
	if pos < 1 || len(lower)-pos-1 < 6 {
		return "", nil, false
	}
	hrp, dataPart := lower[:pos], lower[pos+1:]
	values := make([]int, len(dataPart))
	for i := 0; i < len(dataPart); i++ {
		idx := strings.IndexByte(bech32mCharset, dataPart[i])
		if idx < 0 {
			return "", nil, false
		}
		values[i] = idx
	}
	if bech32mPolymod(append(bech32mHRPExpand(hrp), values...)) != bech32mConst {
		return "", nil, false
	}
	return hrp, values[:len(values)-6], true
}

// Bech32mDecode decodes into (hrp, payload bytes) — the byte-aligned form.
func Bech32mDecode(s string) (hrp string, data []byte, ok bool) {
	h, values, ok := Bech32mDecodeRaw(s)
	if !ok {
		return "", nil, false
	}
	bits := convertBits(values, 5, 8, false)
	if bits == nil {
		return "", nil, false
	}
	return h, intsToBytes(bits), true
}

// --- internals -----------------------------------------------------------

func bech32mPolymod(values []int) int {
	chk := 1
	for _, v := range values {
		b := chk >> 25
		chk = ((chk & 0x1ffffff) << 5) ^ v
		for i := 0; i < 5; i++ {
			if (b>>i)&1 == 1 {
				chk ^= bech32mGen[i]
			}
		}
	}
	return chk
}

func bech32mHRPExpand(hrp string) []int {
	out := make([]int, 0, len(hrp)*2+1)
	for i := 0; i < len(hrp); i++ {
		out = append(out, int(hrp[i])>>5)
	}
	out = append(out, 0)
	for i := 0; i < len(hrp); i++ {
		out = append(out, int(hrp[i])&31)
	}
	return out
}

func bech32mCreateChecksum(hrp string, data []int) []int {
	values := append(append(bech32mHRPExpand(hrp), data...), 0, 0, 0, 0, 0, 0)
	pm := bech32mPolymod(values) ^ bech32mConst
	out := make([]int, 6)
	for i := 0; i < 6; i++ {
		out[i] = (pm >> (5 * (5 - i))) & 31
	}
	return out
}

// convertBits regroups a bit stream; returns nil on failure (a valid empty
// result is a non-nil empty slice).
func convertBits(data []int, from, to int, pad bool) []int {
	acc, bits, maxv := 0, 0, (1<<to)-1
	out := []int{}
	for _, value := range data {
		if value < 0 || (value>>from) != 0 {
			return nil
		}
		acc = (acc << from) | value
		bits += from
		for bits >= to {
			bits -= to
			out = append(out, (acc>>bits)&maxv)
		}
	}
	if pad {
		if bits > 0 {
			out = append(out, (acc<<(to-bits))&maxv)
		}
	} else if bits >= from || ((acc<<(to-bits))&maxv) != 0 {
		return nil
	}
	return out
}

func bytesToInts(b []byte) []int {
	out := make([]int, len(b))
	for i, v := range b {
		out[i] = int(v)
	}
	return out
}

func intsToBytes(vals []int) []byte {
	out := make([]byte, len(vals))
	for i, v := range vals {
		out[i] = byte(v)
	}
	return out
}

File: python/src/hearth/bech32m.py
"""Bech32m encoder/decoder (BIP-350), checksum constant 0x2bc830a3."""

from __future__ import annotations

_CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
_CONST = 0x2BC830A3
_GEN = (0x3B6A57B2, 0x26508E6D, 0x1EA119FA, 0x3D4233DD, 0x2A1462B3)


def encode(hrp: str, data: bytes) -> str:
    values = _convert_bits(list(data), 8, 5, pad=True)
    if values is None:
        raise ValueError("cannot regroup payload bits")
    checksum = _create_checksum(hrp, values)
    return hrp + "1" + "".join(_CHARSET[d] for d in values + checksum)


def decode_raw(s: str) -> tuple[str, list[int]] | None:
    """Verify structure + checksum only; return (hrp, raw 5-bit data groups)."""
    if s != s.lower() and s != s.upper():
        return None
    lower = s.lower()
    pos = lower.rfind("1")
    if pos < 1 or len(lower) - pos - 1 < 6:
        return None
    hrp, data_part = lower[:pos], lower[pos + 1 :]
    values: list[int] = []
    for ch in data_part:
        idx = _CHARSET.find(ch)
        if idx < 0:
            return None
        values.append(idx)
    if _polymod(_hrp_expand(hrp) + values) != _CONST:
        return None
    return hrp, values[:-6]


def decode(s: str) -> tuple[str, bytes] | None:
    """Decode into (hrp, payload bytes) — the byte-aligned form addresses use."""
    raw = decode_raw(s)
    if raw is None:
        return None
    hrp, values = raw
    data = _convert_bits(values, 5, 8, pad=False)
    if data is None:
        return None
    return hrp, bytes(data)


# --- internals -----------------------------------------------------------


def _polymod(values: list[int]) -> int:
    chk = 1
    for v in values:
        b = chk >> 25
        chk = ((chk & 0x1FFFFFF) << 5) ^ v
        for i in range(5):
            if (b >> i) & 1:
                chk ^= _GEN[i]
    return chk


def _hrp_expand(hrp: str) -> list[int]:
    return [ord(c) >> 5 for c in hrp] + [0] + [ord(c) & 31 for c in hrp]


def _create_checksum(hrp: str, data: list[int]) -> list[int]:
    values = _hrp_expand(hrp) + data + [0, 0, 0, 0, 0, 0]
    pm = _polymod(values) ^ _CONST
    return [(pm >> (5 * (5 - i))) & 31 for i in range(6)]


def _convert_bits(data: list[int], frm: int, to: int, pad: bool) -> list[int] | None:
    acc = 0
    bits = 0
    maxv = (1 << to) - 1
    out: list[int] = []
    for value in data:
        if value < 0 or (value >> frm) != 0:
            return None
        acc = (acc << frm) | value
        bits += frm
        while bits >= to:
            bits -= to
            out.append((acc >> bits) & maxv)
    if pad:
        if bits > 0:
            out.append((acc << (to - bits)) & maxv)
    elif bits >= frm or ((acc << (to - bits)) & maxv) != 0:
        return None
    return out
